### DISCUSSION TITLE
How does ResponseType.WAIT work?

### DISCUSSION THREAD

**Discussion by dipplestix** [yasserfarouk/scml#51] [one-shot-open-discussion]
How does using ResponseType.WAIT work? I see negotiations ending after 10-13 wait responses even though it doesn't surpass the 20 second limit for a response? Is there some limit to how many WAITs can be sent to a particular agent? Is it ok to use `time.sleep` to handle this?

**Comment by yasserfarouk** [comment]
How are you suing WAIT? Do you return it directly from your agent's code or is it used within SAOSyncController? The SyncController will not wait more than the number of concurrent negotiations it is controlling. The limit for a  complete round is 20 seconds which gives 10 seconds per negotiator (more or less).

You cannot use `time.sleep` here because the whole simulation runs on a single core (Python's Infamous GIL and our need to run many simulations in parallel for the competition). Using sleep would just lead to wasting the time with no benefit to anyone.

If you let me know what you are trying to do by sending WAIT back, I may be able to help.

Also, are you basing your agent on OneshotAgent or OneshotSyncAgent?

**Reply by dipplestix** [reply]
I'm using OneShotAgent and returning it from agent code.

I want to use WAIT to essentially just stall for time rather than give a response right away for some particular negotiations. 

For example, if I have 3 open negotiations, I want to send (a, b, c) to the first one, (d, e, f) to the second and then see if I can get feedback from them before I send my proposal to the third (but if the difference between state.time and when the negotiation started approaches 20, I can send (g, h, i) still). Right now if I do this, that negotiation just times out.

**Reply by yasserfarouk** [reply]
When you send WAIT to the mechanism, it will just wait and call you again. **Your partner will not have a chance to offer**. 
In your case, you need to wait only on one of the two while saving what you received from them. When you receive the second offer, you now have both and can decide how to respond.

OneShotSyncAgent handles this timing for you. See [this tutorial](http://www.yasserm.com/scml/scml2020docs/tutorials/02.develop_agent_scml2020_oneshot.html#oneshotsyncagent) about it.

**Reply by dipplestix** [reply]
I'm currently doing the following

I have a dictionary `agent.offers` which contains offers I haven't yet responded to. I have a dictionary agent.my_proposals which contains offers I proposed that I haven't received a response to yet

So, I currently have something like this

```
def respond(self, negotiator_id, state, offer):
    if len(self.my_proposals) != 0 and self.time_elapsed[negotiator_id] < 12:
        return ResponseType.WAIT
    else:
        if is_acceptable(offer):
            return  ResponseType.ACCEPT_OFFER
       else:
            self.my_proposals [negotiator_id] = get_proposal(negotiator_id)
            return ResponseType.REJECT_OFFER

def propose(self, negotiator_id: str, state) -> "Outcome":
    if self.offers[negotiator_id] == 'First':
        offer = self.get_first_offer(negotiator_id, state)
        self.offers[negotiator_id] = offer
    else:
        offer = self.my_proposals [negotiator_id]
    return offer
```

Is there a better way to do what I'm trying to do with the first if statement in `respond`?

**Reply by dipplestix** [reply]
From what I can see by running it, the WAIT response just makes the mechanism/world call respond again immediately with the same `neg_id, offer` pair, which would be fine except it seems that after some number of WAIT responses the negotiation is seen as timing out.

**Reply by yasserfarouk** [reply]
Yes that is the expected behavior. Waiting just causes a call back later. Its only effect is to allow other negotiations to take one step. If you are waiting on all your negotiations, none of them can advance which means you will never get the offers you are waiting for and it will timeout eventually. There is an upper limit on the number of waits but even without it, you will see the time out.

That is why we have the OneShotSyncController.

**Reply by yasserfarouk** [reply]
The first if statement seems to cause your agent to wait on all negotiations I think because it evaluates to true in all negotiations first time. 

If you do not like to use OneshotSyncAgent, you can implement something like what the negmas.sap.SAOSyncController is implementing [here](https://github.com/yasserfarouk/negmas/blob/master/negmas/sao/controllers.py). You have to keep track of several pieces to get it working correctly as indicated there.
